### PR TITLE
Implemented ButtonView, and protocol ViewCode

### PR DESCRIPTION
**Pull request by peetorres**
Updated File of SwiftUIPreviews too, handling better little components.

### CODE REVIEW

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Helpers/SwiftUIPreView.swift [hunk 1]
```diff
@@ -36,3 +36,43 @@ import Foundation
...
+    }
+    
+    func updateUIView(_ uiView: View, context: Context) {
+        uiView.setContentHuggingPriority(.defaultHigh, for: .horizontal)
+        uiView.setContentHuggingPriority(.defaultHigh, for: .vertical)
+    }
+}
+
+@available(iOS 13.0, *)
+extension UIViewController {
```
Comment (cipolleschi) [inline solutions/devsprint-riccardo-1/OnboardingChallenge/Helpers/SwiftUIPreView.swift:61]: An important concept when working in a team is that every pr should have a single responsibility. Improving the SwiftUIPreview template was not the responsibility of this PR, which should be focused on implementing the button view. 
This welcomed addendum should go in a separate PR, where we address the possibility to make a ViewController representable in a SwiftUI preview.

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
+//
+//  ViewCode.swift
+//  OnboardingChallenge
+//
+//  Created by Pedro Gabriel on 15/12/21.
+//
+
+import UIKit
```
Comment (cipolleschi) [inline solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift:8]: I see that this file is a way to structure how to write UIView code. I really like this: I'd like to discuss it with the rest of the team. These kinds of changes are useful only if all the team conforms to the same conventions.
 
But, at the moment, we have to remember to make our view conform to it.
Could you think about how we can automate this?

A protocol is an abstraction. Although abstractions can help us organize our work, we shouldn't introduce abstractions we don't need. Do you plan to use the `ViewCode` protocol in other ways other than just have the view conform to it (for example: do you expect to have an array of `ViewCode`)?
If we don't plan to use the protocol in a similar way, it is something we should not add to the codebase.

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+protocol ViewCode where Self: UIView {
```
Comment (cipolleschi) [inline solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift:10]: In this line, the protocol is constrained to `UIView` meaning that only `UIView` can inherit from this protocol.

Could you explain why you make this decision? For example, I could see that a `UIViewController` may benefit from the same methods. And if tomorrow Apple introduces another way to define views with autolayout, they should be able to inherit from the protocol as well.

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    func configureSubviews()
+    func configureSubviewsConstraints()
+    func configureAdditionalBehaviors()
```
Comment (cipolleschi) [inline solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift:13]: What do you expect should go in the `configureAdditionalBehaviors` method?

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Components/ButtonView.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        setupViews()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: Overrides
+    override var intrinsicContentSize: CGSize {
+        return CGSize(width: UIView.noIntrinsicMetric, height: 65)
```
Comment (cipolleschi) [inline solutions/devsprint-riccardo-1/OnboardingChallenge/Components/ButtonView.swift:44]: Great idea to leverage the `intrinsicContentSize`. Given that we are encapsulating a button, shouldn't we return the `button.instrinsicContentSize`?

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Great work so far! 
I left several comments here and there to improve the code and some other considerations. 

Remember to keep the PRs very specific: this pr should contain only the ButtonView code (and related additional support code). The preview part could have been developed in a separate PR.

You correctly followed the request from the task and it's great with some little updates. 
However, we are engineers and not machines. Sometimes we are assigned some tasks but we can implement them in a better way.
In this specific case, we want to create a `ButtonView` that we share across different screens. The button has to work exactly like the classic `UIButton`, but it has some very specific style. 
Can you think about another way to do this task that:
1. Does not require creating a new class
2. Let us work with `UIButton`
3. Allow us to simulate the ButtonView
4. Uses much less code than a class-based approach?

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift (hunk 4 above)
Comment (peetorres) [reply solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift]: This configureAdditionalBehaviors is used to configure things like background, font, or something else that is not addSubviews, or constraints.

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift (hunk 3 above)
Comment (peetorres) [reply solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift]: It's a good question!
But I think that the common Design System conception / usage in this case are elements inside an UIView, so, even you have some button, or some complex UI (compositions), you use UIViews as base class.
Maybe @rdgborges can give another perception about that.
But make sense don't enforce this usage restricted only to UIViews.

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift (hunk 4 above)
Comment (cipolleschi) [reply solutions/devsprint-riccardo-1/OnboardingChallenge/Protocols/ViewCode.swift]: Got it. What could be confusing is that `behaviors` usually means functionalities, like button taps or gestures. In this case, you'd like to wrap some static styling in the `configureAdditionalBehaviors`. For the future, my suggestion is to use more informative names, like `style` and `configureGestures` or `configureAnimations`. In this way, teammates will immediately know what the method will do! 😉

File: solutions/devsprint-riccardo-1/OnboardingChallenge/Components/ButtonView.swift (hunk 5 above)
Comment (cipolleschi) [reply solutions/devsprint-riccardo-1/OnboardingChallenge/Components/ButtonView.swift:46]: This is the only comment standing.

**Comment by cipolleschi** [comment]
Almost done, there is just one comment pending, related to the `intrinsicContentSize`. Once we fix that, we can merge the pr! 💪
